### PR TITLE
Rdar 27397701 synthesized get set typechecking

### PR DESCRIPTION
**Pull request by graydon**
Addresses redundant typechecking of getter/setter function bodies outside of the primary file.
This is responsible for unnecessary quadratic work when doing a multi-file compilation.
PR also includes more scale-testing machinery, and a regression test that uses it.

Resolves [SR-2901](https://bugs.swift.org/browse/SR-2901).
Resolves rdar://27397701

### CODE REVIEW

**Comment by graydon** [comment]
@swift-ci please test

File: include/swift/Basic/Statistic.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_BASIC_STATISTIC_H
+#define SWIFT_BASIC_STATISTIC_H
+
+#include "llvm/ADT/Statistic.h"
+
+#define SWIFT_FUNC_STAT \
```
Comment (huonw) [inline include/swift/Basic/Statistic.h]: Does this need a `do { ... } while (0)` given how often braceless control flow is used, or are macro hacks like that not Swift/LLVM style?

Comment (jrose-apple) [inline include/swift/Basic/Statistic.h]: No, that definitely would be preferred.

File: lib/Sema/CodeSynthesis.cpp [hunk 2]
```diff
@@ -763,17 +767,25 @@ void swift::addTrivialAccessorsToStorage(AbstractStorageDecl *storage,
 
   // Synthesize and type-check the body of the getter.
   synthesizeTrivialGetter(getter, storage, TC);
-  TC.typeCheckDecl(getter, true);
-  TC.typeCheckDecl(getter, false);
+  if (storage->getDeclContext()->isModuleScopeContext()) {
```
Comment (jrose-apple) [inline lib/Sema/CodeSynthesis.cpp]: I don't see why this is the right check. What's different about members? Why can't I just validate accessors at module scope?

**Review by jrose-apple — COMMENTED** [review]
In the future, can you split things like this up into multiple PRs? The statistics-gathering seems separate from the fixes.

**Comment by DougGregor** [comment]
Looks like we can't rely on numpy being around

**Comment by graydon** [comment]
Split scale-test and stats helpers into separate PRs (#5607 and #5608 respectively) that this one depends on. Also fixed numpy and macro definition problems in them.

File: lib/Sema/CodeSynthesis.cpp (hunk 2 above)
Comment (graydon) [inline lib/Sema/CodeSynthesis.cpp]: Without that check, various resilience tests fail when emitting global getters (compiler crashes). I imagine this is due to unconditional emission of those getters, w/o being driven by any other references.

Seems to work when I tighten the condition to "only when in resilient mode", so I'll try that (and add a comment to that extent).

**Comment by graydon** [comment]
Updated to unconditionally typecheck all accessors-with-bodies in decl visitor (that's applied to all SF-local decls) rather than trying to switch behaviour for resilient globals.

PR still depends on #5607 and #5608 to land first.

**Comment by DougGregor** [comment]
This looks fantastic, thanks!

**Comment by graydon** [comment]
@swift-ci please smoke test and merge

**Comment by graydon** [comment]
@swift-ci please smoke test linux and merge

**Comment by graydon** [comment]
@swift-ci please smoke test linux

**Pull request merged**